packages/react-spinbutton/Spec.md

```diff
@@ -2,62 +2,381 @@
+
+`formatter()` and `parser()` allow users to hook into `SpinButton`'s display behavior without having to reimplement features like clamping `value` between `min` and `max` which is required when adding [custom suffixes in v8](https://codepen.io/seanms/pen/QWqpQWp).
+
+`SpinButton` does not allow values less than `min` or greater than `max` when these props are specified. When using the step buttons/arrow keys the value will not step outside the min/max bounds. If a value is typed into `<input>` that is outside the min/max range it will be clamped to the relevant bound, that is values less than `min` will be set to `min` and values greater than `max` will be set to `max`.
+
+Aside from min/max range clamping `SpinButton` does not currently implement any input validation.
+
+No error states are currently implemented.
+
+`SpinButton`'s `onChange` callback is invoked every time a change is committed. A change is committed when:
+
+1. A step button is activated
+2. When a user presses the `Up` or `Down` arrow keys while focused on the component.
+3. When a user presses the `Enter` key while focused on the `<input>`
+4. When the `blur` event is fired from the `<input>`
+
+The `onChange` callback is not invoked while a user is focused on the `<input>` and editing the value of the `SpinButton` directly.
+
+### Component States
 
-- _Component States_
-- _Interaction_
-  - _Keyboard_
-  - _Cursor_
-  - _Touch_
-  - _Screen readers_
+- Normal (no value or valid value, not focused)
+- Focused
+- Disabled
+- Error/Warning (value is invalid in some way). NB: error and validation are not currently supported.
+
+### Interaction
+
+#### Keyboard
+
+- Tab focuses the `<input>` element
+- Up arrow key increments the value by `step` until `max`, if specified.
+- Down arrow key decrements the value by `step` until `min`, if specified.
+- Home key sets the value to `min`, if specified.
+- End key sets the value to `max`, if specified.
+- Tab never focuses the increment or decrement buttons as their functionality is fully available via the keyboard
+
+#### Cursor
+
+- Clicking on the `<input>` focuses it.
+  - Input supports all device/platform mouse interactions (i.e., text selection)
+- Clicking the increment button increases the value by `step` until `max`, if specified.
+- Clicking the decrement button decreases the value by `step` until `min`, if specified.
+
+#### Touch
+
+Same as cursor.
+
+#### Screen Readers
+
+- `<input>` is focusable and editable using standard device/platform interactions.
+- increment and decrement buttons are both focusable and usable with standard device/platform interactions.
```

Comment:
"increment and decrement buttons are both focusable" sounds controversial with these statements from the spec:

> - Tab never focuses the increment or decrement buttons as their functionality is fully available via the keyboard
> Only the `<input>` element of `SpinButton` can be focused via the keyboard.

Should be this updated somehow?

Reply:
Agreed. I'm not sure how the buttons should work with a screen reader and would like @smhigley 's input here.

Reply:
Talked with @smhigley about this and her opinion is that this can really go either way but that she's leaning toward exposing the buttons to screen readers and other ATs because of the many ways people use these tools. For example, mobile screen readers often employ swipes for navigation and a user may expect the button to be focusable in this case.

We still need to do a pass on all our input components to ensure uniformity and we should also do user testing but for now I'm opting to expose these buttons to ATs, knowing that we may decide to reverse that decision in the future.